sno/working_copy/sqlserver.py

```diff
@@ -0,0 +1,553 @@
+                {"schema": self.db_schema},
+            )
+            table_identifiers = ", ".join((self.table_identifier(row[0]) for row in r))
+            if table_identifiers:
+                sess.execute(f"DROP TABLE IF EXISTS {table_identifiers};")
+
+            # Drop schema, unless keep_db_schema_if_possible=True
+            if not keep_db_schema_if_possible:
+                sess.execute(
+                    f"DROP SCHEMA IF EXISTS {self.DB_SCHEMA};",
+                )
+
+    def _create_table_for_dataset(self, sess, dataset):
+        table_spec = sqlserver_adapter.v2_schema_to_sqlserver_spec(
+            dataset.schema, dataset
+        )
+        sess.execute(
+            f"""CREATE TABLE {self.table_identifier(dataset)} ({table_spec});"""
+        )
+
+    def _table_def_for_column_schema(self, col, dataset):
+        if col.data_type == "geometry":
+            # This user-defined Geography type adapts WKB to SQL Server's native geography type.
+            crs_name = col.extra_type_info.get("geometryCRS", None)
+            crs_id = crs_util.get_identifier_int_from_dataset(dataset, crs_name) or 0
+            return sqlalchemy.column(col.name, GeographyType(crs_id))
+        elif col.data_type in ("date", "time", "timestamp"):
+            return sqlalchemy.column(col.name, BaseDateOrTimeType)
+        else:
+            # Don't need to specify type information for other columns at present, since we just pass through the values.
+            return sqlalchemy.column(col.name)
+
+    def _insert_or_replace_into_dataset(self, dataset):
+        pk_col_names = [c.name for c in dataset.schema.pk_columns]
+        non_pk_col_names = [
+            c.name for c in dataset.schema.columns if c.pk_index is None
+        ]
+        return sqlserver_upsert(
+            self._table_def_for_dataset(dataset),
+            index_elements=pk_col_names,
+            set_=non_pk_col_names,
+        )
+
+    def _insert_or_replace_state_table_tree(self, sess, tree_id):
+        r = sess.execute(
+            f"""
+            MERGE {self.SNO_STATE} STA
+            USING (VALUES ('*', 'tree', :value)) AS SRC("table_name", "key", "value")
+            ON SRC."table_name" = STA."table_name" AND SRC."key" = STA."key"
+            WHEN MATCHED THEN
+                UPDATE SET "value" = SRC."value"
+            WHEN NOT MATCHED THEN
+                INSERT ("table_name", "key", "value") VALUES (SRC."table_name", SRC."key", SRC."value");
+            """,
+            {"value": tree_id},
+        )
+        return r.rowcount
+
+    def _write_meta(self, sess, dataset):
+        """Write the title (as a comment) and the CRS. Other metadata is not stored in a PostGIS WC."""
+        self._write_meta_title(sess, dataset)
+
+    def _write_meta_title(self, sess, dataset):
+        """Write the dataset title as a comment on the table."""
+        # TODO - probably need to use sp_addextendedproperty @name=N'MS_Description'
```

Comment:
Not right now - it just means there's currently no place where the dataset `title` is stored in SQL Server working copy, so it can't be changed except using `sno meta set`.

Reply:
Rewrote the comment to be better.